# This is code to replicate the analysis from Alleman et al. "Tandem-Running and Scouting Behavior is Characterized by Up-Regulation of Learning and Memory formation genes within the Ant Brain"
# DESeq2 analysis of T. americanus
# Transcriptome constructed using Trinity and filtered for those contigs that contain an open reading frame
# Expected read counts created using RSEM
# Script by M. Stoldt

####################################################################################################################
## Loading required packages
####################################################################################################################

library(DESeq2)
library(VennDiagram)
library(ggplot2)

####################################################################################################################
## Establishment of directories
####################################################################################################################

# Specify working directory for project.
setwd("/My/Working/Directory/")

# Specify directory where RSEM counts are stored
direc <- "/Directory/RSEM/"        

####################################################################################################################
## Input results generated by RSEM and BLAST
####################################################################################################################

# Read in counting results 
countDirec <- paste(direc, "/RSEM_results", sep="")
resFiles <- list.files(countDirec, include.dirs = TRUE)
countsMatrix <- c()
for (i in 1:length(resFiles)) {
  fileName <- paste(countDirec,paste("/",paste(resFiles[i],"/RSEM.genes.results",sep=""),sep=""), sep="")
  helpData <- read.table(fileName,header=TRUE)
  helpDataOrdered <- helpData[order(helpData$transcript_id.s.),]
  countsMatrix <- cbind(countsMatrix,helpDataOrdered$expected_count)
}
colnames(countsMatrix) <- resFiles
rownames(countsMatrix) <- helpDataOrdered$transcript_id.s.
countsMatrix <- round(countsMatrix)
save(countsMatrix,file="countsMatrix_Tam.Rdata")

# Read in BLAST hits, TAB format.
blast <- read.table("/Directory/BLAST/All_Tam_BLAST_0803.out.tab",header=TRUE,sep="\\t",quote="")

# Create a data frame containing the behavioral groups in the same order as in the countsmatrix
sample_info<-data.frame(group=c("Follower", "Leader", "Scout","Scout", "Leader", "Follower","Leader", "Scout", "Follower","Leader", "Scout", "Follower","Leader", "Follower", "Scout"))
rownames(sample_info)<-names(countsMatrix)

####################################################################################################################
## Additional filtering steps before DESeq2
####################################################################################################################

# Filter out contigs having in less than 4 samples at least 10 counts
countsMatrix_filter1<-countsMatrix[rowSums(countsMatrix>9)>=4,]
# Create a DESeq dataset from the countsmatrix using behavioral group as design
dds <- DESeqDataSetFromMatrix(countsMatrix_filter1, colData=sample_info, design= ~group )
# Run DESeq2 to access information about Cook's distances
dds<-DESeq(dds)

#Extract maximum Cook's distances and average expression
data_cook <- data.frame(gene = rownames(dds), cook = mcols(dds)$maxCooks, average=mcols(dds)$baseMean)

#Plot maximum Cook's distances and average expression
plot(data_cook$average,data_cook$cook, ylim=c(0,45), xlim=c(0,20000), xlab=("Average expression"), ylab=("Maximum Cook's Distance"))
abline(col="red", h=38)

#Extract info on gene ID with Cook's distance below 38
subset_data <- data_cook %>% filter(cook < 38)

#Subset the original countsmatrix by removing the genes filtered out according to Cook's distance >= 38
matrix_cook<-countsMatrix_filter1[match(subset_data$gene, rownames(countsMatrix_filter1)),]
save(matrix_cook,file="matrix_cook_Tam.Rdata")

####################################################################################################################
## DESeq2 Analysis
####################################################################################################################
# Pairwise comparisons using Wald test
####################################################################################################################

# Create DESeq dataset from new filtered matrix
dds <- DESeqDataSetFromMatrix(matrix_cook, colData=sample_info, design= ~group )
# Run DESeq2
dds<-DESeq(dds)

# Acquisition of results
# Compare scouts and leaders and adjust p-values by 0.05
SCvsLE_ds<- results(dds, contrast = c("group", "Scout", "Leader"), alpha=0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_SL<-subset(SCvsLE_ds, padj<0.05&log2FoldChange>0)
annotSL_ds<- blast[match(rownames(padj_SL),blast$Read),]
countSL_ds<-matrix_cook[match(rownames(padj_SL), rownames(matrix_cook)),]
table_SCvsLE <- data.frame("ID"=rownames(padj_SL), "Base Mean"= padj_SL$baseMean, "Log Fold Change"=padj_SL$log2FoldChange, "lfcSE"=padj_SL$lfcSE, "stat"=padj_SL$stat, "p-value"=padj_SL$pvalue, "adjusted p-value"=padj_SL$padj, "blast hit"=annotSL_ds[,2], "read count"=countSL_ds)
write.table(table_SCvsLE,file="DESeq2_ScoutvsLeader_Tam.csv", sep = "\\t", row.names = FALSE)


# Compare scouts and followers and adjust p-values by 0.05
SCvsFO_ds <- results(dds, contrast = c("group", "Scout", "Follower"), alpha=0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_SF<-subset(SCvsFO_ds, padj<0.05&log2FoldChange>0)
annotSF_ds<- blast[match(rownames(padj_SF),blast$Read),]
countSF_ds<-matrix_cook[match(rownames(padj_SF), rownames(matrix_cook)),]
table_SCvsFO <- data.frame("ID"=rownames(padj_SF), "Base Mean"= padj_SF$baseMean, "Log Fold Change"=padj_SF$log2FoldChange, "lfcSE"=padj_SF$lfcSE, "stat"=padj_SF$stat, "p-value"=padj_SF$pvalue, "adjusted p-value"=padj_SF$padj, "blast hit"=annotSF_ds[,2], "read count"=countSF_ds)
write.table(table_SCvsFO,file="DESeq2_ScoutvsFollower_Tam.csv", sep="\\t", row.names = FALSE)


# Compare leaders and scouts and adjust p-values by 0.05
LEvsSC_ds <- results(dds, contrast = c("group", "Leader", "Scout"), alpha = 0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_LS<-subset(LEvsSC_ds, padj<0.05&log2FoldChange>0)
annotLS_ds<- blast[match(rownames(padj_LS),blast$Read),]
countLS_ds<-matrix_cook[match(rownames(padj_LS), rownames(matrix_cook)),]
table_LEvsSC <- data.frame("ID"=rownames(padj_LS), "Base Mean"=padj_LS$baseMean, "Log Fold Change"=padj_LS$log2FoldChange, "lfcSE"=padj_LS$lfcSE, "stat"=padj_LS$stat, "p-value"=padj_LS$pvalue, "adjusted p-value"=padj_LS$padj, "blast hit"=annotLS_ds[,2], "read count"=countLS_ds)
write.table(table_LEvsSC,file="DESeq2_LeadervsScout_Tam.csv", sep="\\t", row.names = FALSE)


# Compare leaders and followers and adjust p-values by 0.05
LEvsFO_ds <- results(dds, contrast = c("group", "Leader", "Follower"), alpha = 0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_LF<-subset(LEvsFO_ds, padj<0.05&log2FoldChange>0)
annotLF_ds<- blast[match(rownames(padj_LF),blast$Read),]
countLF_ds<-matrix_cook[match(rownames(padj_LF), rownames(matrix_cook)),]
table_LEvsFO <- data.frame("ID"=rownames(padj_LF), "Base Mean"=padj_LF$baseMean, "Log Fold Change"=padj_LF$log2FoldChange, "lfcSE"=padj_LF$lfcSE, "stat"=padj_LF$stat, "p-value"=padj_LF$pvalue, "adjusted p-value"=padj_LF$padj, "blast hit"=annotLF_ds[,2], "read count"=countLF_ds)
write.table(table_LEvsFO,file="DESeq2_LeadervsFollower_Tam.csv", sep="\\t", row.names = FALSE)


# Compare followers and scouts and adjust p-values by 0.05
FOvsSC_ds <- results(dds, contrast = c("group", "Follower", "Scout"), alpha = 0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_FS<-subset(FOvsSC_ds, padj<0.05&log2FoldChange>0)
annotFS_ds<- blast[match(rownames(padj_FS),blast$Read),]
countFS_ds<-matrix_cook[match(rownames(padj_FS), rownames(matrix_cook)),]
table_FOvsSC <- data.frame("ID"=rownames(padj_FS), "Base Mean"=padj_FS$baseMean, "Log Fold Change"=padj_FS$log2FoldChange, "lfcSE"=padj_FS$lfcSE, "stat"=padj_FS$stat, "p-value"=padj_FS$pvalue, "adjusted p-value"=padj_FS$padj, "blast hit"=annotFS_ds[,2], "read count"=countFS_ds)
write.table(table_FOvsSC,file="DESeq2_FollowervsScout_Tam.csv", sep="\\t", row.names = FALSE)


# Compare followers and leaders and adjust p-values by 0.05
FOvsLE_ds <- results(dds, contrast = c("group", "Follower", "Leader"), alpha = 0.05)
# Create table with results padj < 0.05 and log fold change > 0 and annotate with BLAST hits and counts
padj_FL<-subset(FOvsLE_ds, padj<0.05&log2FoldChange>0)
annotFL_ds<- blast[match(rownames(padj_FL),blast$Read),]
countFL_ds<-matrix_cook[match(rownames(padj_FL), rownames(matrix_cook)),]
table_FOvsLE <- data.frame("ID"=rownames(padj_FL), "Base Mean"=padj_FL$baseMean, "Log Fold Change"=padj_FL$log2FoldChange, "lfcSE"=padj_FL$lfcSE, "stat"=padj_FL$stat, "p-value"=padj_FL$pvalue, "adjusted p-value"=padj_FL$padj, "blast hit"=annotFL_ds[,2], "read count"=countFL_ds)
write.table(table_FOvsLE,file="DESeq2_FollowervsLeader_Tam.csv", sep="\\t", row.names = FALSE)

####################################################################################################################
## Plots
####################################################################################################################
# Read in the lists where contigs up-regulated in the same behavior are merged
scout <- read.table("deseq2_All_Scout_Tam.csv",header=TRUE,sep="\\t",quote="")
leader <-read.table("deseq2_All_Leader_Tam.csv",header=TRUE,sep="\\t",quote="")
follower <-read.table("deseq2_All_Follower_Tam.csv",header=TRUE,sep="\\t",quote="")

# Combine the lists from above to get the entirety of up-regulated contigs
all<-rbind(scout, leader, follower)
# Remove duplicate entries
all<-unique(all)

####################################################################################################################
## Venn Diagram
####################################################################################################################

tiff("Vennie_Tam.tiff",width = 1280, height = 1230)
venn.plot<-venn.diagram(list(scout$ID, leader$ID, follower$ID), imagetype="tiff",NULL, fontfamily="sans",lwd=c(2.5,2.5,2.5), fill=c("midnightblue", "mediumvioletred", "mediumturquoise"), alpha=c(0.6, 0.6, 0.6), cex=7, cat.fontface="plain" ,cat.fontfamily="sans", category.names = c("Scout", "Leader", "Follower"), cat.cex=c(4,4,4), main="T. americanus", main.fontfamily = "sans", main.fontface = "italic", main.cex=5)
grid.draw(venn.plot)
dev.off()

####################################################################################################################
## PCA plot
####################################################################################################################

# Create DESeqTransform object using rlog
rlog_dds<-rlog(dds, blind=FALSE)

# Plot all data with groups labelled by color and shaped by colony
plotData1<-plotPCA(rlog_dds, intgroup=c("group", "colony"), returnData=TRUE)
percentVar1 <- round(100 * attr(plotData1, "percentVar"))
ggplot(plotData1, aes(x = PC1, y = PC2, color= group.1, shape=colony)) +
  geom_point(size =4, stroke=1) +
  xlab(paste0("PC 1: ", percentVar1[1], "% variance")) +
  ylab(paste0("PC 2: ", percentVar1[2], "% variance")) +
  coord_fixed()+
  scale_color_manual(values=c("midnightblue", "mediumvioletred", "mediumturquoise"), labels=c("Scout", "Leader", "Follower"))+scale_shape_manual(values=c(0:8))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))+
  labs( color = "Group", shape="Colony")

# Extract only the part of the DESeq2 dataset matching to the contig names of the up-regulated contigs
dds_degs<-dds[match(all$ID, rownames(dds))]

# Create DESeqTransform object using rlog
rlog_degs<-rlog(dds_degs, blind=FALSE)

# Plot only the up-regulated contigs with groups labelled by color and shaped by colony
plotData2<-plotPCA(rlog_degs, intgroup=c("group", "colony"), returnData=TRUE)
percentVar2 <- round(100 * attr(plotData2, "percentVar"))
ggplot(plotData2, aes(x = PC1, y = PC2, color= group.1, shape=colony)) +
  geom_point(size =4, stroke=1) +
  xlab(paste0("PC 1: ", percentVar2[1], "% variance")) +
  ylab(paste0("PC 2: ", percentVar2[2], "% variance")) +
  coord_fixed()+
  scale_color_manual(values=c("midnightblue", "mediumvioletred", "mediumturquoise"), labels=c("Scout", "Leader", "Follower"))+scale_shape_manual(values=c(0:8))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))+
  labs( color = "Group", shape="Colony")


